### PR TITLE
Demand workflow

### PR DESCRIPTION
**Pull request by hazemakhalek**
# Closes # 79

## Changes proposed in this Pull Request
This PR downloads the energy balances from UNSD, filters them by the chosen base year and country and processes the data to the sectors and carriers included in the model. The main output is: energy_totals for any country of the world for any historic year.

Following PR will project the energy_balances to the future (planning_horizons) year.

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by hazemakhalek** [comment]
@energyLS @davide-f we have a new package here: py7zr. Should I issue a pypsa-earth PR for this?

**Comment by davide-f** [comment]
Yeah, please :)
Add a Pr to add the dependency

**Comment by hazemakhalek** [comment]
@davide-f, will do

**Comment by hazemakhalek** [comment]
@energyLS , will do. Also finishing up the projection feature. Will ping you when ready

**Comment by hazemakhalek** [comment]
@davide-f I issued the PR for the new dependency.

**Comment by davide-f** [comment]
I'm updating the submodule to the latest version of the repo: https://github.com/pypsa-meets-earth/pypsa-earth-sec/pull/228

That should fix the issue once complete. I hope no breaking changes have occurred

File: config.default.yaml [hunk 1]
```diff
@@ -39,6 +39,10 @@ clustering_options:
 
 countries: ['MA']
 
+demand_data:
+  update_data: true
```
Comment (energyLS) [inline config.default.yaml:43]: Add short explaination: what does this mean?

File: data/demand/efficiency_gains copy.csv [hunk 2]
```diff
@@ -0,0 +1,2 @@
+,total residential space,total residential water,electricity residential,total services space,total services water,total rail,electricity rail,total domestic aviation,total international aviation,total domestic navigation,total international navigation,services electricity,agriculture electricity,agriculture oil,residential oil,residential biomass,residential gas,agriculture biomass,services oil,services biomass,services gas,total road ev,total road fcev,total road ice,total navigation oil,total navigation hydrogen
```
Comment (energyLS) [inline data/demand/efficiency_gains copy.csv:1]: Is this file intended to be here? named "...copy"?

File: data/demand/growth_factors copy.csv [hunk 3]
```diff
@@ -0,0 +1,2 @@
+,total residential space,total residential water,electricity residential,total services space,total services water,total road,total rail,electricity rail,total domestic aviation,total international aviation,total domestic navigation,total international navigation,services electricity,agriculture electricity,agriculture oil,residential oil,residential biomass,residential gas,agriculture biomass,services oil,services biomass,services gas
```
Comment (energyLS) [inline data/demand/growth_factors copy.csv:1]: Is this file intended to be here? named "...copy"?

File: data/energy_totals_XX_2030.csv [hunk 4]
```diff
@@ -0,0 +1,2 @@
+,agriculture biomass,agriculture electricity,agriculture oil,electricity rail,electricity residential,residential biomass,residential gas,residential oil,services biomass,services electricity,services gas,services oil,total domestic aviation,total domestic navigation,total international aviation,total international navigation,total rail,total residential space,total residential water,total road,total services space,total services water,district heat share,electricity residential water,electricity residential space,electricity services space,electricity services water
```
Comment (energyLS) [inline data/energy_totals_XX_2030.csv:1]: Is this file intended to be here?

File: data/energy_totals_XX_2030_.csv [hunk 5]
```diff
@@ -0,0 +1,2 @@
+,agriculture biomass,agriculture electricity,agriculture oil,electricity rail,electricity residential,residential biomass,residential gas,residential oil,services biomass,services electricity,services gas,services oil,total domestic aviation,total domestic navigation,total international aviation,total international navigation,total rail,total residential space,total residential water,total road,total services space,total services water,district heat share,electricity residential water,electricity residential space,electricity services space,electricity services water
```
Comment (energyLS) [inline data/energy_totals_XX_2030_.csv:1]: Is this file intended to be here?

File: data/energy_totals_XX_2038.csv [hunk 6]
```diff
@@ -0,0 +1,2 @@
+,agriculture biomass,agriculture electricity,agriculture oil,electricity rail,electricity residential,residential biomass,residential gas,residential heat biomass,residential heat gas,residential heat oil,residential oil,services biomass,services electricity,services gas,services oil,total domestic aviation,total domestic navigation,total international aviation,total international navigation,total rail,total residential space,total residential water,total road,total services space,total services water,electricity residential space,electricity residential water,district heat share,electricity services space,electricity services water
```
Comment (energyLS) [inline data/energy_totals_XX_2038.csv:1]: Is this file intended to be here? This is 2038 specific output, but not required as input I think.

File: scripts/build_base_energy_totals.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                    ]
+                    .Quantity_TWh.sum()
+                    .round(4)
+                )
+                energy_totals_base.at[country, "residential oil"] = (
+                    df_sector[df_sector.Commodity.isin(oil_fuels)]
+                    .Quantity_TWh.sum()
+                    .round(4)
+                )
+                energy_totals_base.at[country, "residential biomass"] = (
```
Comment (energyLS) [inline scripts/build_base_energy_totals.py:86]: Could be a loop

File: scripts/build_base_energy_totals.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    df = df.loc[df["country"] != "not found"]
+
+    # Convert country column that contains lists for some country names that are identified with more than one country.
+    df["country"] = df["country"].astype(str)
+
+    # Remove all iso2 conversions for some country names that are identified with more than one country.
+    df = df[~df.country.str.contains(",", na=False)].reset_index(drop=True)
+
+    # Create a dictionary with all the conversion factors from ktons or m3 to TWh based on https://unstats.un.org/unsd/energy/balance/2014/05.pdf
+    fuels_conv_toTWh = {
```
Comment (energyLS) [inline scripts/build_base_energy_totals.py:299]: Maybe move to config. Although this is nothing that should be changed...

**Review by energyLS — APPROVED** [review]
First testing shows: 2030 works, 2040 throws error, 2050 works but has the same values as 2030. What is the issue here?

File: scripts/build_base_energy_totals.py (hunk 8 above)
Comment (hazemakhalek) [reply scripts/build_base_energy_totals.py:299]: I thought about it, but I think the config is long enough. This will make it longer unnecessarily. We can add separate files for this and the colors and so on once we have some time.

File: scripts/build_base_energy_totals.py (hunk 7 above)
Comment (hazemakhalek) [reply scripts/build_base_energy_totals.py:86]: Could be, but honestly will be tedious to debug. I would vote to keep it like this as long as we have only one data source (unsd).

File: scripts/prepare_sector_network.py [hunk 9]
```diff
@@ -1296,24 +1296,22 @@ def add_industry(n, costs):
...
-            factor = (
-                1
-                - industrial_demand.loc[loads_i, "current electricity"].sum()
-                / n.loads_t.p_set[loads_i].sum().sum()
-            )
-            n.loads_t.p_set[loads_i] *= factor
-            industrial_elec = industrial_demand["current electricity"].apply(
-                lambda frac: frac / 8760
-            )
+    # if not snakemake.config["custom_data"]["elec_demand"]:
```
[the remark this answers is not in the document]
Comment (hazemakhalek) [reply scripts/prepare_sector_network.py:1299]: because we now use the unsd data for residential electricity demand that doesn't have this issue to begin with

**Comment by hazemakhalek** [comment]
> First testing shows: 2030 works, 2040 throws error, 2050 works but has the same values as 2030. What is the issue here?

"2030 and 2050 are similar": Any time horizon you choose will be based on the parameters of the new csv files (growth factors, efficiency gains and fuel shares). These are not scenario specific, so the user needs to change them when the change the year to to complete the scenario definition. 
"2040 throwing an error": Maybe this is an issue of the costs file not existing for 2040, or do you mean the script itself fails?
